omas/omas_plot.py

```diff
@@ -265,6 +265,64 @@ def gas_filter(label, which_gas):
     return include
 
 
+def gas_arrow(ods, r, z, direction=None, snap_to=numpy.pi/4.0, ax=None, color=None, pad=1.0, **kw):
+    """
+    Draws an arrow pointing in from the gas valve
+    :param ods: ODS instance
+
+    :param r: float
+        R position of gas injector (m)
+
+    :param z: float
+        Z position of gas injector (m)
+
+    :param direction: float
+        Direction of injection (radians, COCOS 1/11)
+
+    :param snap_to: float
+        Snap direction angle to nearest value. Set snap to pi/4 to snap to 0, pi/4, pi/2, 3pi/4, etc. No in-between.
+
+    :param ax: Axes instance to plot on
+
+    :param color: matplotlib color specification
+
+    :param pad: float
+        Padding between arrow tip and specified (r,z)
+    """
+
+    def pick_direction():
+        """Guesses the direction for the arrow (from injector toward machine) in case you don't know"""
+        dr = ods['equilibrium']['time_slice'][0]['global_quantities']['magnetic_axis']['r'] - r
+        dz = ods['equilibrium']['time_slice'][0]['global_quantities']['magnetic_axis']['z'] - z
+        theta = -numpy.arctan2(dz, dr)  # negative is to make it COCOS1/11 compliant
```

Comment:
@orso82 Does ODS apply signs or conversions when I enter data if it's in a different COCOS? I thought I had to save it in COCOS11.

Reply:
Should all internal OMAS operations be COCOS agnostic, or should we fix them to be COCOS 11?

Until now I have coded OMAS so that the ODSs can be in any COCOS. However, we could restrict ourselves and say that the internal OMAS representation is COCOS 11.

The point is that we need to evaluate what are the benefits/usefulness of the generality versus the cost...

Reply:
The schema actually specifies directions for some angles. For example:

gas_injection.pipe[:].exit_position.phi |   | FLT_0D | rad | Toroidal angle (oriented counter-clockwise when viewing from above)
-- | -- | -- | -- | --

So, maybe it has to be COCOS 11 in order to be a valid IDS. I think it would be less confusing (and maybe more legal) to enforce a COCOS internally and just transform on input/output.

Reply:
@eldond When entering data into an ODS, you should determine what COCOS the data is in, then use the cocos_environment for setting the ODS values.  OMAS will then take care of the necessary conversions from the data's COCOS to the ODS's COCOS.  See [OMFITgeqdsk.to_omas](https://github.com/gafusion/OMFIT-source/blob/fa11186517acb631182f0ef207983f30e2a67974/omfit/classes/omfit_eqdsk.py#L711-L845), for example. A valid IDS needs to be in COCOS 11.  Right now, the ODS is free to be in any COCOS.  If it's not in 11, it could be easily transformed to that when written to IMAS.

@orso82 raises a good point though.  If we're going to allow an ODS to have any COCOS, then any internal operations should be coded agnostically, as discussed [here](https://docs.google.com/document/d/1-efimTbI55SjxL_yE_GKSmV4GEvdzai7mAj5UYLLUXw/edit) .  That said, I'm not sure I see the point to maintaining that generality.  We have the ability to read the data out of an ODS in any COCOS.  There's no reason it can't live inside of the data structure in COCOS 11.

Reply:
@bclyons12 Thanks. Within the scope of these hardware descriptions, where I have to gather data from different sources, I think it's easier if I just write it all in COCOS11 and apply negative signs as needed to make that happen. I have to watch out for unit conversions anyway, so I might as well do the signs at the same time.

In a different context, where you already have a consistent dataset, it sounds like it would be easier to do as you say and tell OMAS you're loading data from a different COCOS.

Reply:
@eldond IMAS is COCOS 11, and the documentation is consistent with it (which is good).

My point is that this:
```python
theta = -numpy.arctan2(dz, dr)  # negative is to make it COCOS1/11 compliant
```
is not much simpler than this
```python
theta = cocos_transform(3, ods.cocos)['POL']*numpy.arctan2(dz, dr)  #transform POLoidal angle from ctr-clockwise cartesian system (COCOS 3) to ODS COCOS
```
and the later one is totally general. Also I like the fact that it is explicit.

Reply:
I am calculating the angle right here. I am not obtaining it from some other data set that already had a COCOS. I get to choose the COCOS for my calculation because there was no COCOS before. I chose COCOS11 to make it simpler; this way no more transforms are needed. 

My calculations are natively COCOS11 because I put in the `-`. I commented it to explain the native COCOS of my calculation.

Given all that, should I then put `cocos_transform(11, ods.cocos)*` in front of `-numpy.arctan2(dz, dr)`, to be safe in case `ods.cocos` is not 11?

Reply:
@eldond I wouldn't do it that way, but there is the danger that ods.cocosin (not ods.cocos) somehow got permanently set to something that isn't 11.  The idea is that you should only ever change ods.cocosin using "with cocos_environment," though right now I don't believe that is enforced anywhere.  

There are two options then:
 1. Define the angle as you do. I think -arctan(z/r) is general enough that it wouldn't need a cocos_transform multiplier in front of it.  Then set the angle in the ods inside a " with cocos_environment(ods,cocosin=11):"
 2. @orso82 and/or I change the ODS so that ods.cocos is always 11 and ods.cocosin/ods.cocosout can only be changed using a "with cocos_environment".  If you have your data in 11 already, then you wouldn't need to declare the environment before populating the ODS.